Remove dependency on IRGenModule of some linkage computations

huonw opened the pull request:
This is designed to allow TBDGen to reuse this for computing information about all the various symbols it touches.

huonw committed:
[IRGen] Remove unnecessary IRGenModule arguments.

huonw committed:
[IRGen] Expose LinkInfo publicly, like LinkEntity.

huonw committed:
[IRGen] Remove IGM argument, factor out conformance fragility logic.

huonw committed:
[IRGen] Compute LinkInfo without an IGM.

slavapestov commented on include/swift/AST/ProtocolConformance.h:
```diff
@@ -266,7 +266,11 @@ class alignas(1 << DeclAlignInBits) ProtocolConformance {
+
+  /// Returns true if this conformance should be serialized, based on the
+  /// type/protocol involved in it.
+  bool shouldBeSerialized() const;
```
Nitpick: the AST doesn't "know" about serialization. Maybe call this hasFixedLayout to be consistent with types

huonw replied:
Okay.

slavapestov commented on include/swift/IRGen/Linking.h:
```diff
@@ -10,22 +10,31 @@
 namespace irgen {
 class IRGenModule;
 
+class UniversalLinkageInfo {
```
Does anything that reference IRGenModile belong in this file at all?

huonw replied:
I could flip the helper constructor of this around to `IRGenModule::getUniversalLinkageInfo()` or something, I suppose, and remove the `LinkInfo::get` helper overload, but I'm not entirely sure that's worth it. (Also, `isAvailableExternally` seemed like it would become far more ugly to become independent of `IRGenModule`: the argument would be replaced by a `DeclContext*` meant represent the module-scope of the `IRGenModule`.)

slavapestov reviewed (commented):
Nice cleanup!

huonw commented:
@swift-ci Please smoke test.

I updated the PR with some of the feedback, and also added a bit of an unfortunate hack to the first commit. I'm curious if it seems sensible.

```swift
    // A hidden-external conformance doesn't make sense: we can't refer to
    // it. Thus, the only way to end up in that situation is if the linkage is
    // actually public.
    return linkage == SILLinkage::HiddenExternal ? SILLinkage::PublicExternal
                                                 : linkage;
```

huonw commented:
@swift-ci Please smoke test.

huonw commented:
@swift-ci Please smoke test

huonw commented:
https://github.com/apple/swift-lldb/pull/177

@swift-ci Please smoke test

slavapestov commented:
Hidden external means the symbol is defined in a different translation unit of the same module. Think two files that reference each other's internal symbols. This can totally happen -- something else must have gone wrong.
